### DISCUSSION TITLE
checkpoint every module in a different ckpt file

### DISCUSSION THREAD

**Discussion by malfonsoarquimea** [Lightning-AI/pytorch-lightning#10840] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi!
I am currently working on a project where I would like to checkpoint my model in separated pieces. 
My model has a backbone composed by:

- a backbone, which is also composed by 3 modules
- several heads, each one being a module
I would like to save one ckpt with the backbone and one ckpt per head. I understand that I should create a custom callback inheriting from ModelCheckpoint and then modifying on_save_checkpoint, I am not really aware of how to do it.
on_save_checkpoint is defined as: 

![image](https://user-images.githubusercontent.com/95293295/144058369-9930d133-734a-46cc-9413-9c663a4d6ac9.png)
 
Another solution would be to modify my lightning module to load the ckpt when the training ends as a dict and then save each subpart as a ckpt using torch.save(), but I understand that this solution is much less elegant.

Any suggestions? Thanks in advance!

**Comment by rohitgr7** [comment, marked as answer]
I'd suggest using [checkpoint_io](https://pytorch-lightning.readthedocs.io/en/latest/advanced/checkpoint_io.html) plugin for your use-case.

**Reply by malfonsoarquimea** [reply]
First of all, thanks very much for your answer!
I have some doubts regarding the checkpoint_io plugin
If I understood correctly, I should modify the method save_checkpoint() on my custom version of CheckpointIO(). My doubt (probably a noob doubt) is how this will interact with the ModelCheckpoint callback I am using in my training loop. I am a bit confused about callbacks and hooks and how they interact with each other, so if you could provide some insight on this particular topic or redirect me to some source of knowledge to understand how the callbacks and hooks actually work I would be really happy!

PD: Do you have any example of a custom CheckpointIO?
Thanks in advance!

**Reply by rohitgr7** [reply]
the checkpoint-related hooks in model_checkpoint are usually used to manipulate the checkpoint itself, more like what to checkpoint and add/delete/modify stuff in it. The `CheckpointIO` is built for users to give them the ability to modify how the checkpoint is saved and how it's being loaded.

for your use-case:
```py
from pytorch_lightning.plugins.io import CheckpointIO

class MyCheckpointIO(CheckpointIO):
    def save_checkpoint(self, checkpoint, path, ...):
        # break the checkpoint['state_dict'] and store it separately

    def load_checkpoint(path, ...):
        # load the checkpoint and reconstruct it


custom_checkpoint_io = MyCheckpointIO()
trainer = Trainer(... plugins=[custom_checkpoint_io])
```
you can check the default one here for reference: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/io/torch_plugin.py

we also released a blog post recently that you can check: https://devblog.pytorchlightning.ai/lightning-checkpointio-plugin-f8663d8bc5e8

**Reply by malfonsoarquimea** [reply]
Thank you very much! I found really useful both things, the blog and the example. I managed to successfully save and load my ckpts as I wanted. Thanks very much!
Is there a way to update the state of the discussion to solved?

**Reply by rohitgr7** [reply]
> Is there a way to update the state of the discussion to solved?

you can click on `Mark as answer` above in the first comment.

Also if possible can you share that piece of code if its possible for others looking for it in future?

**Reply by malfonsoarquimea** [reply]
For sure, I only implemented a dummy version where my model is structured in the base model and last layer. The state dict of the checkpoint contains the following keys
![image](https://user-images.githubusercontent.com/95293295/144216676-0104f7de-a2fd-4c00-b666-555bd53a87d2.png)

With that in mind, I created a custom IO plugin to save 2 ckpts, one with all the layers in the base model (contains the word "base" in the layer name) and the ones in the last layer (does not contain the word "base" in the layer name) 
![image](https://user-images.githubusercontent.com/95293295/144216100-4643698a-6584-4a54-b6bd-f2ab38b5cab6.png)

and the output of my training script is:
![image](https://user-images.githubusercontent.com/95293295/144412379-16ebb165-519e-465c-af51-eefa6d1efb84.png)
